d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -368,10 +425,52 @@ private void onData(ResourceUpdate data)
         _log.debug("Received resource update data equal to the current data. Will not perform the update.");
         return;
       }
-      _data = data;
+      if (!isEmptyData(data))
+      {
+        // null value guard to avoid overwriting the property with null
+        _data = data;
+      }
+      if (_data == null)
+      {
+        _data = initEmptyData();
```

Comment:
Actually here the ResourceUpdate could have a right version (the one added on line 238 and 271), which should be kept. You can pass it to initEmptyData. 

private ResourceUpdate initEmptyData(String version)
    {
      switch (_type)
      {
        case NODE:
          return new NodeUpdate(version != null ? version : VERSION_FOR_NULL_DATA, null);
        case D2_URI_MAP:
          return new D2URIMapUpdate(version != null ? version : VERSION_FOR_NULL_DATA, null);
        case UNKNOWN:
        default:
          throw new AssertionError("should never be here");
      }
    }

Reply:
Hi, @bohhyang Actually I don't think there we should use the `right version` u mentioned. As this logic is `initEmptyData` fro `_data`, which means the original `_data` in `subscriber` is null, and as `data` is invalid, so we need create `an empty` data for `_data`. So I think ur former proposal sounds better. What do u think

Reply:
@PapaCharlie brought up this idea that when the data is null/empty, but the version is available, we don't drop the version. 
I think it could be beneficial for the client to keep the version, it can be used for debugging, such as adding it to the log at line 228 about which version is invalid, so that we can compare the version with the observer log to find the particular bad data. Also, it's generally good to keep the version in _data, so that we know which version the client has received.